File: ETrade/src/main/java/com/mthree/etrade/service/StockService.java
package com.mthree.etrade.service;

import com.mthree.etrade.model.Stock;
import com.mthree.etrade.model.StockPrice;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

public interface StockService {

    Stock getStock(String symbol);

    List<Stock> getAllStocks();

    BigDecimal getCurrentPrice(String symbol);

    void updateStock(Stock stock);

    void addStock(Stock stock);

    void deleteStock(String symbol);

    List<Stock> searchStock(String keyword);

    List<StockPrice> getStockHistory(String symbol, LocalDate startDate, LocalDate endDate);
}

File: frontend/src/services/stockService.js
import apiClient from './api';

// Stock service methods
const stockService = {
    // Get all stocks
    getAllStocks: async () => {
        try {
            const response = await apiClient.get('/api/stocks');
            return response.data;
        } catch (error) {
            console.error('Error fetching stocks:', error);
            throw error;
        }
    },

    // Get stock by symbol
    getStockBySymbol: async (symbol) => {
        try {
            const response = await apiClient.get(`/api/stocks/${symbol}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching stock ${symbol}:`, error);
            throw error;
        }
    },

    // Search stocks by keyword
    searchStocks: async (keyword) => {
        try {
            const response = await apiClient.get(`/api/stocks/search/${keyword}`);
            return response.data;
        } catch (error) {
            console.error(`Error searching stocks with keyword ${keyword}:`, error);
            throw error;
        }
    },

    // Get current price for a stock
    getCurrentPrice: async (symbol) => {
        try {
            const response = await apiClient.get(`/api/stocks/price/${symbol}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching price for stock ${symbol}:`, error);
            throw error;
        }
    },

    // Get historical data for a stock
    getStockHistory: async (symbol, startDate, endDate) => {
        try {
            const response = await apiClient.get(`/api/stocks/history/${symbol}/${startDate}&${endDate}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching history for stock ${symbol}:`, error);
            throw error;
        }
    }
};

export default stockService;